### PR TITLE
Decredtx, Helper, tests

### PR DESCRIPTION
**Pull request by drmoog**
This commit has in it all the decred-related edits I've made, with summa's master merged in and conflict-resolved including the name and location changes in tx.py

Next for this is to work out the values for the commented-out tests (sighashing on decredtx), which are WIP from previous PR, #68.

### CODE REVIEW

File: riemann/tests/helpers.py [hunk 1]
```diff
@@ -831,8 +832,72 @@
             }
         }
 
+DCR2 = {
```
Comment (prestwich) [inline riemann/tests/helpers.py]: Add a line with a link to this tx
e.g. 
```
#  http://explorer.dcrdata.org/api/tx/hex/9b23d239a9507e652741ee4135384277b56442302e333df0ca83c71f20d580a8
```

File: riemann/tests/tx/test_tx.py [hunk 2]
```diff
@@ -1253,17 +1332,17 @@ def test_init_errors(self):
...
-        #         version=self.version,
-        #         tx_ins=[self.tx_in] * 2,
-        #         tx_outs=[self.tx_out],
-        #         lock_time=self.lock_time,
-        #         expiry=self.expiry,
-        #         tx_witnesses=[self.witness])
-        #
-        # self.assertIn('Witness and TxIn lists must be same length. ',
-        #               str(context.exception))
+        with self.assertRaises(ValueError) as context:
```
Comment (prestwich) [inline riemann/tests/tx/test_tx.py]: Can we improve this with an `attr_assert` function?

File: riemann/tests/tx/test_tx.py [hunk 3]
```diff
@@ -1345,6 +1470,37 @@ def test_witness_hash(self):
             transaction.witness_hash(),
             helpers.DCR['ser']['witnesses'][0]['hash'])
 
+    def test_get_script_code(self):
+        transaction = tx.DecredTx(
```
Comment (prestwich) [inline riemann/tests/tx/test_tx.py]: Can we improve this with a dict on the self object created in setup, and `tx.DecredTx(**self.tx_args)` like in zcash?

File: riemann/tests/tx/test_tx.py [hunk 4]
```diff
@@ -1345,6 +1470,37 @@ def test_witness_hash(self):
...
+            lock_time=self.lock_time,
+            expiry=self.expiry,
+            tx_witnesses=[self.witness])
+
+        self.assertEqual(
+            transaction._get_script_code(index=0),
+            helpers.DCR['ser']['witnesses'][0]['redeem_script'])
+
+    def test_script_code_fail(self):
+        self.witness = tx.DecredInputWitness(
```
Comment (prestwich) [inline riemann/tests/tx/test_tx.py]: Why `self.` here?

File: riemann/tests/tx/test_tx.py [hunk 5]
```diff
@@ -1345,6 +1470,37 @@ def test_witness_hash(self):
...
+            value=self.witness_value,
+            height=self.height,
+            index=self.witness_index,
+            stack_script=self.stack_script,
+            redeem_script=b'')
+
+        transaction = tx.DecredTx(
```
Comment (prestwich) [inline riemann/tests/tx/test_tx.py]: See earlier comment about `**self.tx_args`

File: riemann/tests/tx/test_tx.py [hunk 6]
```diff
@@ -1440,34 +1597,56 @@ def test_sighash(self):
 
         tx_witnesses = [tx_witness_0, tx_witness_1, tx_witness_2]
 
-        tx.DecredTx(
+        t = tx.DecredTx(
```
Comment (prestwich) [inline riemann/tests/tx/test_tx.py]: `**self.tx_args**

File: riemann/tests/tx/test_tx.py [hunk 7]
```diff
@@ -1440,34 +1597,56 @@ def test_sighash(self):
...
             version=helpers.DCR1['ser']['version'],
             tx_ins=tx_ins,
             tx_outs=tx_outs,
             lock_time=helpers.DCR1['ser']['locktime'],
             expiry=helpers.DCR1['ser']['expiry'],
             tx_witnesses=tx_witnesses)
 
+        # test sighash_all:
```
Comment (prestwich) [inline riemann/tests/tx/test_tx.py]: Are we 100% sure the sighash all test vector is correct?

File: riemann/tx/tx.py [hunk 8]
```diff
@@ -442,10 +442,6 @@ def from_bytes(InputWitness, byte_string):
             items.append(item)
         return InputWitness(items)
 
-    def copy(self, stack=None):
```
Comment (prestwich) [inline riemann/tx/tx.py]: Where did this go?

File: riemann/tx/tx.py [hunk 9]
```diff
@@ -1218,7 +1310,10 @@ def sighash_single(self, index, script=None,
 
         try:
             copy_tx_outs = copy_tx.tx_outs[:index + 1]
-            copy_tx_outs = [TxOut(value=b'\xff' * 8, output_script=b'')
+            copy_tx_outs = [DecredTxOut(
```
Comment (prestwich) [inline riemann/tx/tx.py]: Good catch :+1:

File: tox.ini [hunk 10]
```diff
@@ -21,7 +21,7 @@ commands =
     --exclude test_tx.py \
     riemann
   pytest \
-    -q \
+    -s \
```
[the remark this answers is not in the document]
Comment (drmoog) [reply tox.ini]: √

File: riemann/tests/helpers.py (hunk 1 above)
Comment (drmoog) [reply riemann/tests/helpers.py]: √

File: riemann/tests/tx/test_tx.py (hunk 2 above)
Comment (drmoog) [reply riemann/tests/tx/test_tx.py]: Elegant, on it

File: riemann/tests/tx/test_tx.py (hunk 4 above)
Comment (drmoog) [reply riemann/tests/tx/test_tx.py]: ( mild shame )

File: riemann/tx/tx.py (hunk 8 above)
Comment (drmoog) [reply riemann/tx/tx.py]: done R-U-N-N-O-F-T it appears, have sent the bloodhounds

**Comment by prestwich** [comment]
@drmoog so uh, I have a big restructure in the works that will make it easier for me to review things like this. buuuut it means I'm going to close this because merge conflict from the void beyond the stars. It's not like I was getting to it anyway >.>

**Pull request closed without merging**